awt/toolkit-0.95.cs

```diff
@@ -5225,137 +5225,175 @@ internal string getNativeClipboardFormatName(long format)
-                    }
-                    else
-                    {
-                        // don't know what to do with it...
-                        obj.SetData(transferable.getTransferData(flavor));
-                    }
-                }
-                catch (java.io.IOException e)
-                {
-                    if (!(flavor.isMimeTypeEqual(java.awt.datatransfer.DataFlavor.javaJVMLocalObjectMimeType) &&
-                          e is java.io.NotSerializableException))
-                    {
-                        e.printStackTrace();
-                    }
-                }
-            }
-            return obj;
-        }
+        internal Map translateFromClipboard(IDataObject data)
+        {
+            java.awt.datatransfer.FlavorTable defaultFlavorMap = (java.awt.datatransfer.FlavorTable)java.awt.datatransfer.SystemFlavorMap.getDefaultFlavorMap();
+            Map/*<DataFlavor,object>*/ map = new HashMap();
+            if (data == null)
+            {
+                return map;
+            }
+            string[] formats = data.GetFormats();
+            if (formats != null && formats.Length > 0)
+            {
+                long[] longFormats = getClipboardFormatCodes(formats);
+                Map /*<DataFlavor,long>*/ flavorMap = getFlavorsForFormats(longFormats, defaultFlavorMap);
+                java.awt.datatransfer.DataFlavor[] flavors =
+                    (java.awt.datatransfer.DataFlavor[])
+                    (flavorMap.keySet().toArray(new java.awt.datatransfer.DataFlavor[0]));
+                for (int i = 0; i < flavors.Length; i++)
+                {
+                    java.awt.datatransfer.DataFlavor df = flavors[i];
+                    long format = ((java.lang.Long)flavorMap.get(df)).longValue();
+                    string stringFormat = getNativeClipboardFormatName(format);
+                    if (stringFormat == null) continue; // clipboard format is not registered in Windows system
+                    object translatedData = null;
+                    if (df.isFlavorJavaFileListType())
+                    {
+                        object formatData = data.GetData(stringFormat);
+                        if (formatData == null) continue; // no data for that format
+
+                        // translate string[] into java.util.List<java.io.File>
+                        string[] nativeFileList = (string[])formatData;
+                        List fileList = new ArrayList(nativeFileList.Length);
+                        for (int j = 0; j < nativeFileList.Length; j++)
+                        {
+                            java.io.File file = new java.io.File(nativeFileList[i]);
+                            fileList.add(file);
+                        }
+                        translatedData = fileList;
+                    }
+                    else if (java.awt.datatransfer.DataFlavor.imageFlavor.equals(df))
+                    {

```

Comment:
I guess we ignore images but I think a short comment, to understand why we skip it, would be welcome.

Reply:
I saw IKVM crashed when I tested with images in clipboard. This probable depends on type of bitmap because Windows supports multiple bitmap formats for clipboard. Support of images in clipboard is not required for Atoms. Implementation of this is outside of scope of this change.
Unfortunately the above text is not good comment for code but I think it will be available in pull request history.